src/Conversion/ONNXToTOSA/Math/Elementwise.cpp

```diff
@@ -0,0 +1,71 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//===---------------- Elementwise.cpp - Elementwise Op --------------------===//
+//
+// Copyright (c) 2022 Advanced Micro Devices, Inc.
+//
+// =============================================================================
+//
+// This file lowers ONNX element-wise operators to TOSA dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#include "src/Conversion/ONNXToTOSA/ONNXToTOSACommon.hpp"
+
+using namespace mlir;
+
+namespace onnx_mlir {
+
+namespace {
+
+template <typename ONNXOpT, typename TOSAOpT>
+class ONNXUnaryOpLoweringToTOSA : public OpConversionPattern<ONNXOpT> {
+public:
+  using OpConversionPattern<ONNXOpT>::OpConversionPattern;
+  using OpAdaptor = typename ONNXOpT::Adaptor;
+  LogicalResult matchAndRewrite(ONNXOpT op, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    rewriter.replaceOpWithNewOp<TOSAOpT>(op, op.getType(), adaptor.X());
+    return success();
+  }
+};
+
+class ONNXReluOpLoweringToTOSA : public OpConversionPattern<ONNXReluOp> {
+public:
+  using OpConversionPattern::OpConversionPattern;
+  LogicalResult matchAndRewrite(ONNXReluOp op, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+
+    Value input = adaptor.X();
+
+    // Quantized types are not supported right now (in type conversion).
+    // Once they are, the input should be rescaled for quantized types. (TBD)
+    // Maps to `tosa.clamp` which has both int and fp limits.
+    rewriter.replaceOpWithNewOp<tosa::ClampOp>(op, op.getType(), input,
+        rewriter.getI64IntegerAttr(0),
+        rewriter.getI64IntegerAttr(std::numeric_limits<int32_t>::max()),
+        rewriter.getF32FloatAttr(0.0f),
+        rewriter.getF32FloatAttr(std::numeric_limits<float>::max()));
+    return success();
+  }
+};
+
+} // namespace
+
+void populateLoweringONNXElementwiseOpToTOSAPattern(ConversionTarget &target,
+    RewritePatternSet &patterns, TypeConverter &typeConverter,
+    MLIRContext *ctx) {
+  patterns.insert<ONNXReluOpLoweringToTOSA>(typeConverter, ctx);
+
+#define INSERT_UNARY_PATTERN(ONNXOp, TOSAOp)                                   \
+  target.addIllegalOp<ONNXOp>();                                               \
+  patterns.insert<ONNXUnaryOpLoweringToTOSA<ONNXOp, TOSAOp>>(                  \
+      typeConverter, ctx);
+  INSERT_UNARY_PATTERN(ONNXNegOp, tosa::NegateOp)
+  INSERT_UNARY_PATTERN(ONNXFloorOp, tosa::FloorOp)
```

Comment:
A subtle distinction between Floor and Negate is that Floor is defined in the TOSA spec only for fp types. Negate is defined for both integer and fp. 
https://www.mlplatform.org/tosa/tosa_spec.html#_floor
https://www.mlplatform.org/tosa/tosa_spec.html#_negate

This is currently noted as a bug in the TorchToTosa path as well - I remembered to check since I'd implemented that piece myself.

Reply:
We could either not use the macro mechanism and have a dedicated pattern, or provide an explicit template specialization here. In either case, it seems appropriate to use the same `notifyMatchFailure` approach on integer types to floor for now.

Reply:
We use a few macros here and there, generally we try to avoid them and use templated functions or some other C++ mechanisms.

Reply:
Thank you for the quick response! By my understanding `onnx.Floor` will never use an integer type in `onnx-mlir` https://github.com/onnx/onnx/blob/main/docs/Operators.md#Floor. Nevertheless, I have updated the type converter. Hope this is slightly clearer.